design/namespaced-roles.md

```diff
@@ -0,0 +1,299 @@
+   - Supports empty values - A role might be created at any level and not necessarily assigned to a group.
+   - Example: `role_v1:/realm1/tenantA/clientId/groupB/roleName`, `role_v1:/realm1/tenantA/clientId//roleName`
+
+**RoleName**: Finally, the name of the role that is being defined.
+   - Doesn't support empty values - You need to know the name of the role that is being created.
+   - Example: `role_v1:/realm1/tenantA/clientId/groupB/roleName`
+
+A possible free-form role identifier could be allowed for resource-server specific roles.
+
+The initial fixed part would be kept in order to identify the role within a realm, tenant, client, group etc, but instead of a fixed name for the role, it would be a reference to other hierarchical role formed with any random structure.
+
+A possible format would be:
+
+- `role_v1:/realm1/tenant1//group1/frole_v1:/mycompany/resources/department-a-roles/developer`
+
+If the role name is another reference to a free form role, starting with `frole_v1:/`, this role will also be represented in a hierarchical way, making it possible for protocol mappers to show them as such in tokens.
+
+## Use Cases
+
+As previously mentioned, there are a few use cases that we'll want to support and are currently not possible in Keycloak.
+
+### Use Case 1 - Assign a role in the context of a group
+
+Say we want to have the following group composition in our realm (or our tenant, in the future), most likely imported from a federated LDAP:
+
+realm/tenant:tenant1
+ - group:iam
+ - group:devops
+ - group:leadership
+
+Following the current implementation, the users would need to define roles at the realm level and then assign them to these groups. Everyone in those groups would inherit the roles.
+
+But for a client, these groups may have business meaning or reflect organizational structure within their company, they aren't just a mean to group users together with the same roles, but rather fully functional teams that users can be part of with different roles.
+
+So UserA would be a `manager` in `group:iam`, but UserB could be a `developer`in `group:iam`.
+
+Namespaced roles would be specified when created, and loosely assigned based on the namespace, so let's say we have the following roles:
+
+`role_v1:/realm1/tenant1//iam/manager`
+`role_v1:/realm1/tenant1//iam/developer`
+`role_v1:/realm1/tenant1//devops/developer`
+`role_v1:/realm1/tenant1//devops/devops_role`
+
+Now these roles are assigned directly to `UserA` and `UserB` so that:
+
+- `UserA` has the `role_v1:/realm1/tenant1//iam/manager` and the `role_v1:/realm1/tenant1//devops/developer` roles.
+- `UserB` has the `role_v1:/realm1/tenant1//iam/developer` and the `role_v1:/realm1/tenant1//devops/developer` roles.
+
+Also, we should be able to still support giving everyone in a group an inherited role, such as:
+
+- `Group:devops` has the `role_v1:/realm1/tenant1//devops/devops_role` role.
+
+#### Representation in the tokens
```

Comment:
Would like to see some alternatives on how roles can be represented in a token. For example options to flatten them if that makes sense. For example a client may not understand the concept of role namespaces, but would only want the roles for a specific namespace flattened.

Also, would be good to touch on how the roles can be represented in the current way they are being done in tokens with regards to realm and client roles.